### PR TITLE
feat(iot-dev): Add support for PnP Model Id

### PR DESCRIPTION
**Pull request by bikamani**

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+package com.microsoft.azure.sdk.iot.device;
+
+import lombok.Getter;
+import lombok.Setter;
+
+public final class ClientOptions
+{
+    /**
+     * The Plug and Play Model Id associated with the device identity.
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java:12]: I am calling this the `digital twin model id`

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientConfig.java [hunk 2]
```diff
@@ -38,6 +39,10 @@
     private boolean useWebsocket;
     private ProxySettings proxySettings;
 
+    @Getter
+    @Setter(AccessLevel.PROTECTED)
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientConfig.java:43]: Isn't lombok so cool! :)

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/net/IotHubUri.java [hunk 3]
```diff
@@ -28,6 +28,9 @@
     /** The API version will be passed as a param in the URI. */
     public static final String API_VERSION = "api-version=" + TransportUtils.IOTHUB_API_VERSION;
 
+    /** The API version will be passed as a param in the URI for PnP. */
+    public static final String PNP_API_VERSION = "api-version=" + TransportUtils.PNP_IOTHUB_API_VERSION;
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/net/IotHubUri.java:32]: I renamed this to API_VERSION_PREVIEW, and added a comment that clarifies that this is for pnp support

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/TransportUtils.java [hunk 4]
```diff
@@ -8,6 +8,7 @@
 public class TransportUtils
 {
     public static String IOTHUB_API_VERSION = "2019-10-01";
+    public static String PNP_IOTHUB_API_VERSION = "2020-05-31-preview";
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/TransportUtils.java:11]: This as "preview" as well

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java [hunk 5]
```diff
@@ -151,7 +151,18 @@ else if (this.config.getAuthenticationType() == DeviceClientConfig.AuthType.X509
                     clientId += "/" + moduleId;
                 }
 
-                this.iotHubUserName = this.config.getIotHubHostname() + "/" + clientId + "/" + API_VERSION + "&" + clientUserAgentIdentifier;
+                String apiVersion;
+                String modelId = this.config.getModelId();
+                if(modelId == null)
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java:156]: `modelId == null || modelId.isEmpty()`

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java [hunk 6]
```diff
@@ -151,7 +151,18 @@ else if (this.config.getAuthenticationType() == DeviceClientConfig.AuthType.X509
                     clientId += "/" + moduleId;
                 }
 
-                this.iotHubUserName = this.config.getIotHubHostname() + "/" + clientId + "/" + API_VERSION + "&" + clientUserAgentIdentifier;
+                String apiVersion;
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java:154]: rename this to serverParam or something similar, this has both api version and model Id (for preview APIs)

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/TransportUtils.java (hunk 4 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/TransportUtils.java:11]: sure can change it

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java [hunk 7]
```diff
@@ -147,6 +148,34 @@ public void constructorSuccess() throws URISyntaxException, IOException
...
+        new Verifications()
+        {
+            {
+                IotHubConnectionString iotHubConnectionString = Deencapsulation.newInstance(IotHubConnectionString.class, connString);
+                times = 1;
+
+                IoTHubConnectionType ioTHubConnectionType = Deencapsulation.getField(client, "ioTHubConnectionType");
+                assertEquals(SINGLE_CLIENT, ioTHubConnectionType);
+
+                TransportClient transportClient = Deencapsulation.getField(client, "transportClient");
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:173]: how is this assertion relevant here?

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java [hunk 8]
```diff
@@ -245,6 +310,28 @@ public void createFromSecurityProviderUsesUriAndDeviceIdAndSavesSecurityProvider
...
+        clientOptions.setModelId("dtmi:Company:device:1");
+
+        //act
+        DeviceClient.createFromSecurityProvider(expectedUri, expectedDeviceId, mockSecurityProvider, expectedProtocol);
+
+        //assert
+        new Verifications()
+        {
+            {
+                //TODO add check for super() call
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:330]: Todo here

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java (hunk 6 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java:154]: I know. I though about a better naming! I would be looking any other suggestions too other than serverParam. :)

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java [hunk 9]
```diff
@@ -147,6 +148,34 @@ public void constructorSuccess() throws URISyntaxException, IOException
...
+        clientOptions.setModelId("dtmi:Company:device:1");
+
+        // act
+        final DeviceClient client = new DeviceClient(connString, protocol, clientOptions);
+
+        // assert
+        new Verifications()
+        {
+            {
+                IotHubConnectionString iotHubConnectionString = Deencapsulation.newInstance(IotHubConnectionString.class, connString);
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:167]: What are we doing with `iotHubConnectionString` here?

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java [hunk 10]
```diff
@@ -222,6 +251,42 @@ public void constructorWithSSLContextSuccess(@Mocked final SSLContext mockedSSLC
...
+        new Verifications()
+        {
+            {
+                IoTHubConnectionType ioTHubConnectionType = Deencapsulation.getField(client, "ioTHubConnectionType");
+                assertEquals(SINGLE_CLIENT, ioTHubConnectionType);
+
+                TransportClient transportClient = Deencapsulation.getField(client, "transportClient");
+                assertNull(transportClient);
+
+                new DeviceClientConfig(mockIotHubConnectionString, mockedSSLContext);
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:285]: Can you explain what we are testing here?

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java (hunk 7 above)
Comment (bikamani) [reply device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:173]: From our legacy tests:  The constructor shall set the transportClient to null.

What do you think is releavant?

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java [hunk 11]
```diff
@@ -308,6 +310,53 @@ public void openEstablishesConnectionUsingCorrectConfig() throws IOException, Tr
...
+        MqttIotHubConnection connection = new MqttIotHubConnection(mockConfig);
+        Deencapsulation.setField(connection, "listener", mockedIotHubListener);
+        connection.open(mockedQueue, mockedScheduledExecutorService);
+
+        final String actualIotHubUserName = Deencapsulation.getField(connection, "iotHubUserName");
+
+        String clientIdentifier = "DeviceClientType=" + URLEncoder.encode(TransportUtils.USER_AGENT_STRING, "UTF-8").replaceAll("\\+", "%20");
+        assertTrue(actualIotHubUserName.contains(iotHubHostName + "/" + deviceId + "/" + PNP_API_VERSION));
+
+        final String actualUserPassword = Deencapsulation.getField(connection, "iotHubUserPassword");
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java:343]: nit: this assertion is not really relevant to this test

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java [hunk 12]
```diff
@@ -308,6 +310,53 @@ public void openEstablishesConnectionUsingCorrectConfig() throws IOException, Tr
         };
     }
 
+    @Test
+    public void openEstablishesConnectionUsingModelIdSetsCorrectPreviewApiVersion() throws IOException, TransportException
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java:314]: Add another test for scenario where you don't set the model Id

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java (hunk 9 above)
Comment (bikamani) [reply device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:167]: good catch, it's unused

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java (hunk 7 above)
Comment (abhipsaMisra) [reply device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:173]: I don't understand why initializing a device client sets the transportClient to null, and how that is relevant to this test case.

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java (hunk 12 above)
Comment (bikamani) [reply device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java:314]: If you look at openEstablishesConnectionUsingCorrectConfig - this does not set model id and makes sure API version is IOTHUB_API_VERSION so it is already covered

Comment (abhipsaMisra) [reply device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java:314]: cool

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java (hunk 1 above)
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java:12]: Add a note that non plug and play users should not set this value

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package com.microsoft.azure.sdk.iot.device;
+
+import lombok.Getter;
+import lombok.Setter;
+
+public final class ClientOptions
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java:9]: Add a description to this class. Something about how it is for constructor time options that won't change for as long as this client exists

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java [hunk 14]
```diff
@@ -173,7 +173,7 @@ public DeviceClient(String connString, TransportClient transportClient) throws U
     public DeviceClient(String connString, IotHubClientProtocol protocol) throws URISyntaxException, IllegalArgumentException
     {
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:174]: Just make this constructor call your new constructor but with a null client options. No need to call commonConstructorVerifications/setup here and in your new constructor

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java [hunk 15]
```diff
@@ -11,6 +11,7 @@
 import com.microsoft.azure.sdk.iot.device.fileupload.FileUpload;
 import com.microsoft.azure.sdk.iot.device.transport.amqps.IoTHubConnectionType;
 import com.microsoft.azure.sdk.iot.provisioning.security.SecurityProvider;
+import com.sun.security.ntlm.Client;
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:14]: What is this import for?

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java (hunk 8 above)
Comment (timtay-microsoft) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:330]: TODO?

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java [hunk 16]
```diff
@@ -42,8 +42,10 @@
     final String iotHubHostName = "test.host.name";
     final String hubName = "test.iothub";
     final String deviceId = "test-deviceId";
+    final String modelId = "dtmi:my:company:namespace;1";
     final String deviceKey = "test-devicekey?&test";
-    final String API_VERSION = Deencapsulation.getField(MqttIotHubConnection.class, "API_VERSION");
+    final String API_VERSION = "?api-version=" + TransportUtils.IOTHUB_API_VERSION;
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java:47]: factor out the "?api-version=" string and reference it in these two places

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java [hunk 17]
```diff
@@ -229,7 +255,26 @@ public DeviceClient(String connString, IotHubClientProtocol protocol, String pub
...
+     *                   and it may be a SAS connection string (format: "HostName=...;DeviceId=...;SharedAccessKey=..."). If
+     *                   this connection string is an x509 connection string, the client will use the provided SSLContext for authentication.
+     * @param protocol the protocol to use when communicating with IotHub
+     * @param sslContext the ssl context that will be used during authentication. If the provided connection string does not contain
+     *                   SAS based credentials, then the sslContext will be used for x509 authentication. If the provided connection string
+     *                   does contain SAS based credentials, the sslContext will still be used during SSL negotiation.
+     * @param clientOptions The options that allow configuration of the device client instance during initialization
+     * @throws URISyntaxException if the hostname in the connection string is not a valid URI
+     */
+    public DeviceClient(String connString, IotHubClientProtocol protocol, SSLContext sslContext, ClientOptions clientOptions) throws URISyntaxException
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:275]: Why can't SSLContext be part of clientOptions?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java [hunk 18]
```diff
@@ -247,7 +292,24 @@ public DeviceClient(String connString, IotHubClientProtocol protocol, SSLContext
...
     public static DeviceClient createFromSecurityProvider(String uri, String deviceId, SecurityProvider securityProvider, IotHubClientProtocol protocol) throws URISyntaxException, IOException
     {
-        return new DeviceClient(uri, deviceId, securityProvider, protocol);
+        return new DeviceClient(uri, deviceId, securityProvider, protocol, null);
+    }
+
+    /**
+     * Creates a device client that uses the provided security provider for authentication.
+     *
+     * @param uri The connection string for iot hub to connect to (format: "yourHubName.azure-devices.net")
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:301]: URI is not connection string. Isn't it hostname only?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientConfig.java (hunk 2 above)
Comment (prmathur-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientConfig.java:43]: why not private?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ModuleClient.java [hunk 19]
```diff
@@ -79,7 +79,7 @@
     public ModuleClient(String connectionString, IotHubClientProtocol protocol) throws ModuleClientException, IllegalArgumentException, UnsupportedOperationException, URISyntaxException
     {
         //Codes_SRS_MODULECLIENT_34_006: [This function shall invoke the super constructor.]
-        super(new IotHubConnectionString(connectionString), protocol, SEND_PERIOD_MILLIS, getReceivePeriod(protocol));
+        super(new IotHubConnectionString(connectionString), protocol, SEND_PERIOD_MILLIS, getReceivePeriod(protocol), null);
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ModuleClient.java:82]: Why does IotHubConnectionString class need client options? config already has it right?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/TransportUtils.java (hunk 4 above)
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/TransportUtils.java:11]: Please add a detail comment as to why are we adding preview API here

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientConfig.java (hunk 2 above)
Comment (vinagesh) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientConfig.java:43]: lombok looks cool. New learnings for me :)

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java (hunk 1 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java:12]: will do

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java (hunk 14 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:174]: Agreed

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java (hunk 15 above)
Comment (bikamani) [reply device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:14]: I'll remove it, it may have been added when I was adding some changes.

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java (hunk 8 above)
Comment (bikamani) [reply device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:330]: I'll take care of this.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java (hunk 17 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:275]: How would that work with other constructors?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java (hunk 18 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:301]: it is hostname. will modify this comment

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java (hunk 5 above)
Comment (vinagesh) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java:156]: Looks like java does not have a single function for this :)

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/TransportUtils.java (hunk 4 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/TransportUtils.java:11]: added

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java (hunk 6 above)
Comment (vinagesh) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java:154]: An option to consider might be - Define two variabes - apiVersion, modelIdParam and initialize both inside the if-else. When you do str concatenation later, it should just work (if it works like c#)
this.iotHubUserName = this.config.getIotHubHostname() + "/" + clientId + "/?api-version=" + apiVersion + modelIdParam

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java [hunk 20]
```diff
@@ -147,6 +148,34 @@ public void constructorSuccess() throws URISyntaxException, IOException
         };
     }
 
+    @Test
+    public void constructorWithClientOptionsSuccess() throws URISyntaxException, IOException
+    {
+        // arrange
+        final String connString = "HostName=iothub.device.com;CredentialType=SharedAccessKey;CredentialScope=Device;deviceId=testdevice;SharedAccessKey=adjkl234j52=;";
```
Comment (vinagesh) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:155]: is this a real connection string or fake?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java (hunk 17 above)
Comment (vinagesh) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:275]: Is it possible to just have 2 constructors and maybe SSLContext can be set to null if user does not want to use it?

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java (hunk 20 above)
Comment (bikamani) [reply device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:155]: It is fake may be never changed in our legacy test code. I am removing these tests as they seem unnecessary after rethinking. But this is good to address in follow up PR so I'll keep a note of this.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java (hunk 6 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java]: I like your suggestion. I also need to check how it works in case where modelId is not part of username. Currently I have it renamed as per Abhipsa's comment

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java [hunk 21]
```diff
@@ -229,15 +247,32 @@ public DeviceClient(String connString, IotHubClientProtocol protocol, String pub
...
+        super(new IotHubConnectionString(connString), protocol, sslContext, SEND_PERIOD_MILLIS, getReceivePeriod(protocol), clientOptions);
         commonConstructorVerifications();
         commonConstructorSetup();
     }
 
     /**
      * Creates a device client that uses the provided security provider for authentication.
      *
-     * @param uri The connection string for iot hub to connect to (format: "yourHubName.azure-devices.net")
+     * @param uri The hostname of iot hub to connect to (format: "yourHubName.azure-devices.net")
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:275]: the hostname of _the_

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.Setter;
+
+/**
+ * Options that allow configuration of the device client instance during initialization.
+ */
+public final class ClientOptions
+{
+    /**
+     * The Digital Twin Model Id associated with the device identity.
+     * Non plug and play users should not set this value
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java:16]: FYI - we do not want to use the term "plug and play" in any public comments/ APIs.

Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java:16]: I am confused, are you suggesting to document this as Digital Twin users?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java (hunk 17 above)
Comment (prmathur-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:265]: As this is a new API, I don't see reason why you can't do that.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java [hunk 23]
```diff
@@ -42,7 +42,7 @@
     private static final String SSL_PREFIX = "ssl://";
     private static final String SSL_PORT_SUFFIX = ":8883";
 
-    private static final String API_VERSION = "?api-version=" + TransportUtils.IOTHUB_API_VERSION;
+    private static String ModelIdParam = "digital-twin-model-id";
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java:45]: shouldn't this be final?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java (hunk 22 above)
Comment (abhipsaMisra) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java:16]: I am not sure, `digital twin users` could potentially confuse users, since there is a different service called "digital twins". 
@prmathur-microsoft - any suggestions on which term to use here?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java (hunk 23 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java:45]: agreed, will address this

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java (hunk 22 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java:16]: exactly, that is the reason why I am asking.

File: device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java [hunk 24]
```diff
@@ -129,7 +129,7 @@ public void constructorSuccess() throws URISyntaxException, IOException
         final IotHubClientProtocol protocol = IotHubClientProtocol.AMQPS;
 
         // act
-        final DeviceClient client = new DeviceClient(connString, protocol);
+        final DeviceClient client = new DeviceClient(connString, protocol, (ClientOptions) null);
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:132]: Why did we have to modify existing e2e?

Comment (prmathur-microsoft) [reply device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java:132]: this is ut. nm

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java (hunk 22 above)
Comment (prmathur-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ClientOptions.java:16]: I would leave it at model id and revisit this when we have a better story around this.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java (hunk 17 above)
Comment (bikamani) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java:265]: We have resolved this offline and I will be sending out a PR for these changes.
